component/src/sct/view.rs

```diff
@@ -0,0 +1,127 @@
+        tracing::debug!(?index, ?sct_block_anchor, "writing epoch anchor");
+
+        self.put(state_key::epoch_anchor_by_index(index), sct_block_anchor);
+        self.put_proto(state_key::epoch_anchor_lookup(sct_block_anchor), index);
+    }
+
+    async fn write_sct(
+        &mut self,
+        height: u64,
+        sct: tct::Tree,
+        block_root: tct::builder::block::Root,
+        epoch_root: Option<tct::builder::epoch::Root>,
+    ) {
+        self.set_sct_anchor(height, sct.root());
+
+        self.set_sct_block_anchor(height, block_root);
+
+        if let Some(epoch_root) = epoch_root {
+            let index = self.epoch().await.expect("epoch must be set").index;
+            self.set_sct_epoch_anchor(index, epoch_root);
+        }
+
+        self.stub_put_state_commitment_tree(&sct);
+    }
+}
+
+impl<T: StateWrite + ?Sized> StateWriteExt for T {}
+
+#[async_trait]
+pub trait StateReadExt: StateRead {
+    async fn stub_state_commitment_tree(&self) -> tct::Tree {
+        match self
+            .nonconsensus_get_raw(state_key::stub_state_commitment_tree().as_bytes())
+            .await
+            .unwrap()
+        {
+            Some(bytes) => bincode::deserialize(&bytes).unwrap(),
+            None => tct::Tree::new(),
+        }
+    }
+
+    async fn anchor_by_height(&self, height: u64) -> Result<Option<tct::Root>> {
+        self.get(&state_key::anchor_by_height(height)).await
+    }
+
+    async fn check_claimed_anchor(&self, anchor: tct::Root) -> Result<()> {
+        if anchor.is_empty() {
+            return Ok(());
+        }
+
+        if let Some(anchor_height) = self
+            .get_proto::<u64>(&state_key::anchor_lookup(anchor))
+            .await?
+        {
+            tracing::debug!(?anchor, ?anchor_height, "anchor is valid");
+            Ok(())
+        } else {
+            Err(anyhow!(
+                "provided anchor {} is not a valid SCT root",
+                anchor
+            ))
+        }
+    }
+
+    async fn finish_sct_block(
+        &self,
+        compact_block: &mut CompactBlock,
+        state_commitment_tree: &mut tct::Tree,
+    ) {
+        let height = compact_block.height;
+
+        // Close the block in the TCT
+        let block_root = state_commitment_tree
+            .end_block()
+            .expect("ending a block in the state commitment tree can never fail");
+
+        // Put the block root in the compact block
+        compact_block.block_root = block_root;
+
+        // If the block ends an epoch, also close the epoch in the TCT
+        if self.epoch().await.unwrap().is_epoch_end(height) {
+            tracing::debug!(?height, "end of epoch");
+
+            let epoch_root = state_commitment_tree
+                .end_epoch()
+                .expect("ending an epoch in the state commitment tree can never fail");
+
+            // Put the epoch root in the compact block
+            compact_block.epoch_root = Some(epoch_root);
+        }
```

Comment:
It feels a little uncomfy that this is duplicating decide-if-we're-in-an-epoch-end logic, since the goal of the refactoring here is to have a single decision point for that at the app level.  Instead, I think we should either lift this logic up to the app (it's not really sct-specific, since it touches both the sct and the compact block), or have a parameter passed in that instructs `finish_sct_block` about whether or not to end the epoch.